### PR TITLE
Support for scraping multiple mysqld hosts

### PR DESCRIPTION
**Pull request by MattiasAng**
This is a continuation on the work done by [samitpal](https://github.com/samitpal) in #504 to address PR comments and add `module` parameter support.

### CODE REVIEW

**Comment by SuperQ** [comment]
Nice, thanks for picking this up.

**Comment by SuperQ** [comment]
CC @syadmind who is working on similar support for the postgres_exporter.

We're using an INI style file here, rather than yaml, since it's more MySQL style.

**Comment by MattiasAng** [comment]
I am wondering if having the `target` for default config is really necessary if users specifically need to enable support for exporting multiple instances? To me it feels like it only adds complexity with no real benefit.

**Review by sysadmind — COMMENTED** [review]
I think that this PR is functionally correct. I would prefer to see this align with the `/probe` pattern as outlined by the prometheus docs, but otherwise, I think things look good in general.

**Comment by MattiasAng** [comment]
@sysadmind Thanks, I will have a look at refactoring in to `/probe` endpoint but can't promise when yet as I have to prioritise other work.

**Comment by MattiasAng** [comment]
@sysadmind  @SuperQ Probe endpoint is added now. Config could probably be handled a bit better as @sysadmind has done on postgres_exporter but due to time constraint I only refactored the parsing.

I have yet to update / clean up documentation so ignore that for now, would appreciate any feedback on the implementation.

**Comment by SuperQ** [comment]
Nice, thanks. I'll take a look at this soon.

File: probe.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+		registry.MustRegister(collector.New(ctx, dsn, metrics, filteredScrapers, logger))
+
+		if err != nil {
+			probeSuccessGauge.Set(0)
+			probeDurationGauge.Set(time.Since(start).Seconds())
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		duration := time.Since(start).Seconds()
```
Comment (sysadmind) [inline probe.go:84]: I'm not sure that the duration is accurate because this will be set before the actual metrics are collected below.

File: mysqld_exporter.go [hunk 2]
```diff
@@ -104,8 +103,7 @@ var scrapers = map[collector.Scraper]bool{
 	collector.ScrapeReplicaHost{}:                         false,
 }
 
-func parseMycnf(config interface{}) (string, error) {
-	var dsn string
+func parseMycnf(config interface{}) (dsn string, err error) {
```
Comment (sysadmind) [inline mysqld_exporter.go:106]: typically we try and stay away from named returns. I would remove the names here.

**Review by sysadmind — CHANGES_REQUESTED** [review]

File: probe.go (hunk 1 above)
Comment (MattiasAng) [reply probe.go]: You are right, and also quite unnecessary since the collector handles timing as well.
Removed it.

**Comment by MattiasAng** [comment]
Updated documentation and some overall cleanup from the previous implementation.

Do you want me to add anything to the changelog or do you handle that on release?

**Review by sysadmind — APPROVED** [review]
LGTM @SuperQ should be able to comment on the Changelog. I'm not familiar with the release process for this project.

**Comment by SuperQ** [comment]
So I mentioned removing ENV var support in https://github.com/prometheus/mysqld_exporter/issues/306.

I'm wondering if we should include some basic command line / ENV replacement for this.

Maybe if we just had `MYSQLD_EXPORTER_PASSWORD`.

This could help issues like https://github.com/prometheus/mysqld_exporter/issues/458, we could properly escape passwords in the connection string to the MySQL driver.

I think we still want to support the sidecar use case easily, without a config file for credentials.

**Comment by MattiasAng** [comment]
> Maybe if we just had `MYSQLD_EXPORTER_PASSWORD`.

If we want to support sidecar I would say we would probably need to have both `MYSQLD_EXPORTER_PASSWORD` and `MYSQLD_EXPORTER_USER`, since I think (or hope?) that most users have a separate user for the exporter.

My proposal if we want to have this replacement for `DATA_SOURCE_NAME` is that we only implement it for the `/metrics` endpoint and that the environment variables takes precedence if they are set.

**Comment by SuperQ** [comment]
For configuration that's not in the file, we prefer command line arguments over env vars. The only exception needed is for secrets.

So rather than `MYSQLD_EXPORTER_USER`, we would want `--mysqld.username`.

File: README.md [hunk 3]
```diff
@@ -112,6 +147,13 @@ web.listen-address                         | Address to listen on for web interf
 web.telemetry-path                         | Path under which to expose metrics.
 version                                    | Print the version information.
 
+### Environment Variables
+Name                                       | Description
+-------------------------------------------|--------------------------------------------------------------------------------------------------
+MYSQLD_EXPORTER_HOST                       | Hostname to be used for connecting to MySQL Server
+MYSQLD_EXPORTER_USER                       | Username to be used for connecting to MySQL Server
```
Comment (SuperQ) [inline README.md:154]: These two should be command line flags, not env vars.

**Comment by MattiasAng** [comment]
> For configuration that's not in the file, we prefer command line arguments over env vars. The only exception needed is for secrets.
> 
> So rather than `MYSQLD_EXPORTER_USER`, we would want `--mysqld.username`.

Thanks for commenting so fast.

While I understand your line of thinking, I don't fully agree with going for that approach.

If a user still needs to set an environment variable (for the password) I feel that we are better off having all of these as environment variables. The only use-case where you don't need to set an environment variable in that case is for servers that do not require a password.

So in short, I would rather keep it as-is now, but if you insist it should be quite easy to fix by bringing in the arguments to `ReloadConfig`.

**Comment by SuperQ** [comment]
I do insist, This is the require design pattern for exporters. Any non-secret options should be command line arguments to follow our design pattern.

File: mysqld_exporter.go [hunk 4]
```diff
@@ -59,11 +54,21 @@ var (
 		"config.my-cnf",
 		"Path to .my.cnf file to read MySQL credentials from.",
 	).Default(path.Join(os.Getenv("HOME"), ".my.cnf")).String()
+	mysqldHost = kingpin.Flag(
+		"mysqld.host",
+		"Hostname to use for connecting to MySQL",
+	).Default("localhost").String()
```
Comment (SuperQ) [inline mysqld_exporter.go:60]: Shouldn't this be `mysql.address`? With a default of `localhost:3306`? 

Then `ReloadConfig()` can use `net.SplitHostPort()` to get the host and port.

Comment (MattiasAng) [reply mysqld_exporter.go]: Good point, for some reason I was thinking that I had added a `port` flag as well.

Decided against `net.SplitHostPort()` since it doesn't allow port to be optional, which would cause errors if a user specifies the `mysqld.address` without a port.

File: config/config.go [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+		certPairs = append(certPairs, keypair)
+		tlsCfg.Certificates = certPairs
+	}
+	tlsCfg.InsecureSkipVerify = m.TlsInsecureSkipVerify
+	mysql.RegisterTLSConfig("custom", &tlsCfg)
+	return nil
+}
+
+func parseAddress(address string) (string, string) {
+	target := strings.Split(address, ":")
```
Comment (SuperQ) [inline config/config.go:227]: I'd prefer to use Go `net.SplitHostPort()`. This will break with IPv6 addresses.

https://pkg.go.dev/net#SplitHostPort

**Review by SuperQ — CHANGES_REQUESTED** [review]

**Review by SuperQ — APPROVED** [review]
Awesome, thanks so much for all the work and requested changes.

File: Makefile [hunk 6]
```diff
@@ -22,8 +22,8 @@ STATICCHECK_IGNORE =
 
 DOCKER_IMAGE_NAME ?= mysqld-exporter
 
-test-docker:
-	@echo ">> testing docker image"
+test-docker-single-exporter:
```
Comment (roidelapluie) [inline Makefile:25]:
```suggestion
.PHONY: test-docker-single-exporter
test-docker-single-exporter:
```

File: README.md [hunk 7]
```diff
@@ -112,6 +149,15 @@ web.listen-address                         | Address to listen on for web interf
 web.telemetry-path                         | Path under which to expose metrics.
 version                                    | Print the version information.
 
+### Environment Variables
+Name                                       | Description
+-------------------------------------------|--------------------------------------------------------------------------------------------------
+MYSQLD_EXPORTER_PASSWORD                   | Password to be used for connecting to MySQL Server
```
Comment (roidelapluie) [inline README.md:155]: we generally avoid passwords as env variables

File: config/config.go [hunk 8]
```diff
@@ -0,0 +1,229 @@
+// Copyright 2018 The Prometheus Authors
```
Comment (roidelapluie) [inline config/config.go:1]:
```suggestion
// Copyright 2022 The Prometheus Authors
```

File: config/config.go [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+	if cfg, err = ini.LoadSources(
+		opts,
+		[]byte("[client]\npassword = ${MYSQLD_EXPORTER_PASSWORD}\n"),
+		filename,
+	); err != nil {
+		return fmt.Errorf("failed to load %s", filename)
+	}
+
+	if host, port, err = net.SplitHostPort(mysqldAddress); err != nil {
+		return fmt.Errorf("failed to parse address: %s", err)
```
Comment (roidelapluie) [inline config/config.go:106]:
```suggestion
		return fmt.Errorf("failed to parse address: %w", err)
```

File: config/config.go [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+			configReloadSeconds.SetToCurrentTime()
+		}
+	}()
+
+	if cfg, err = ini.LoadSources(
+		opts,
+		[]byte("[client]\npassword = ${MYSQLD_EXPORTER_PASSWORD}\n"),
+		filename,
+	); err != nil {
+		return fmt.Errorf("failed to load %s", filename)
```
Comment (roidelapluie) [inline config/config.go:102]: return the error as well

File: config/config.go [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+		}
+	}
+
+	cfg.ValueMapper = os.ExpandEnv
+	config := &Config{}
+	m := make(map[string]MySqlConfig)
+	for _, sec := range cfg.Sections() {
+		sectionName := sec.Name()
+
+		if sectionName == "DEFAULT" {
```
Comment (roidelapluie) [inline config/config.go:127]: we should skip client section as well

File: config/config.go [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+	} else {
+		if host, port, err = net.SplitHostPort(target); err != nil {
+			return dsn, fmt.Errorf("failed to parse target: %s", err)
+		}
+		dsn = fmt.Sprintf("%s:%s@tcp(%s:%s)/", user, password, host, port)
+	}
+
+	if m.SslCa != "" {
+		if tlsErr := m.CustomizeTLS(); tlsErr != nil {
+			tlsErr = fmt.Errorf("failed to register a custom TLS configuration for mysql dsn: %s", tlsErr)
```
Comment (roidelapluie) [inline config/config.go:195]:
```suggestion
			tlsErr = fmt.Errorf("failed to register a custom TLS configuration for mysql dsn: %w", tlsErr)
```

File: config/config.go [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+		if socket != "" {
+			dsn = fmt.Sprintf("%s:%s@unix(%s)/", user, password, socket)
+		} else {
+			dsn = fmt.Sprintf("%s:%s@tcp(%s:%d)/", user, password, host, port)
+		}
+	} else {
+		if host, port, err = net.SplitHostPort(target); err != nil {
+			return dsn, fmt.Errorf("failed to parse target: %s", err)
+		}
+		dsn = fmt.Sprintf("%s:%s@tcp(%s:%s)/", user, password, host, port)
```
Comment (roidelapluie) [inline config/config.go:190]: what if a password contains : ? shouldn't we build this more carefully?

File: config/config.go [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+	if ok := caBundle.AppendCertsFromPEM(pemCA); ok {
+		tlsCfg.RootCAs = caBundle
+	} else {
+		return fmt.Errorf("failed parse pem-encoded CA certificates from %s", m.SslCa)
+	}
+	if m.SslCert != "" && m.SslKey != "" {
+		certPairs := make([]tls.Certificate, 0, 1)
+		keypair, err := tls.LoadX509KeyPair(m.SslCert, m.SslKey)
+		if err != nil {
+			return fmt.Errorf("failed to parse pem-encoded SSL cert %s or SSL key %s: %s",
```
Comment (roidelapluie) [inline config/config.go:220]:
```suggestion
			return fmt.Errorf("failed to parse pem-encoded SSL cert %s or SSL key %s: %w",
```

File: config/config_test.go [hunk 15]
```diff
@@ -0,0 +1,173 @@
+// Copyright 2018 The Prometheus Authors
```
Comment (roidelapluie) [inline config/config_test.go:1]:
```suggestion
// Copyright 2022 The Prometheus Authors
```

File: mysqld_exporter.go [hunk 16]
```diff
@@ -59,11 +54,21 @@ var (
 		"config.my-cnf",
 		"Path to .my.cnf file to read MySQL credentials from.",
 	).Default(path.Join(os.Getenv("HOME"), ".my.cnf")).String()
+	mysqldAddress = kingpin.Flag(
+		"mysqld.address",
+		"Address to use for connecting to MySQL",
+	).Default("localhost:3306").String()
```
Comment (roidelapluie) [inline mysqld_exporter.go:60]: should we just use localhost and connect via unix socket by default?

File: probe.go [hunk 17]
```diff
@@ -0,0 +1,76 @@
+// Copyright 2018 The Prometheus Authors
```
Comment (roidelapluie) [inline probe.go:1]:
```suggestion
// Copyright 2022 The Prometheus Authors
```

File: README.md (hunk 7 above)
Comment (SuperQ) [reply README.md:155]: This is the one place where I am allowing this, as I want a way to avoid needing a config file for simple deployments.

File: config/config.go (hunk 13 above)
Comment (SuperQ) [reply config/config.go:190]: This has been a long-standing issue with various Go SQL drivers. They all seem to suffer from this parsing problem.

There's https://pkg.go.dev/github.com/go-sql-driver/mysql?utm_source=godoc#Config which should probably be used here.

File: mysqld_exporter.go (hunk 16 above)
Comment (SuperQ) [reply mysqld_exporter.go:60]: The current default is localhost, I'd rather stick with that.

File: config/config.go (hunk 11 above)
Comment (MattiasAng) [reply config/config.go:127]: That would break configuration for those that are currently using it with a config file rather than `DATA_SOURCE_NAME`, are we OK with that?

Comment (SuperQ) [reply config/config.go:127]: No, I think we need to keep this so it continues to work as-is.

File: config/config.go (hunk 13 above)
Comment (MattiasAng) [reply config/config.go:190]: I am wondering if we couldn't leave this for another issue to rework? This PR is already adding quite a lot.

It will require quite a fair amount of work to pass that along to the Exporter and parse parameters, and my time is quite limited right now.

**Review by SuperQ — APPROVED** [review]

**Comment by roidelapluie** [comment]
LGTM

**Comment by SuperQ** [comment]
Let's do this!

File: config/config.go (hunk 13 above)
Comment (SuperQ) [reply config/config.go:190]: That's fine. Let's work on it next.

**Comment by yinshurman** [comment]
so how to configure multiple hosts and ports ? 
when i set host and port in different section of mysql.cnf , just the one in [client] makes effect.
and when i only set user and password in client ,and configure prometheus.yml file like this:
`
  - job_name: "mysql" # to get metrics  about mysql 
    params:
        # not required. default `client`
        auth_module: [client]
    static_configs:
      - targets: 
        # all mysql hostname to monitor 
        # - '10.203.4.249:10003'
        # - '10.203.4.249:10016'
        - '172.22.0.5:3306'
        - '172.22.0.11:3307'
        - '172.22.0.12:3308'
    relabel_configs:
      - source_labels: [__address__]
        target_label: __param_target
      - source_labels: [__param_target]
        target_label: instance 
      - target_label: __address__ 
        # the mysqld exporter host:port
        replacement: 172.22.0.6:9104
`
the mysqld exporter reports error information like this :
`ts=2022-10-14T02:05:48.270Z caller=exporter.go:149 level=error msg="Error pinging mysqld" err="dial tcp 127.0.0.1:3306: connect: connection refused"`


so , it seems must set host and port in some where other than prometheus.yml , if not ,the defautl  127.0.0.1:3306 is used .

any one can make it clear  where to set the different hosts and ports ? 
please help ! thanks very much !

**Comment by Suomea** [comment]
@yinshurman looks like not release yet?

**Comment by Suomea** [comment]
> The documentation is missing that `metrics_path` needs to be changed to `/probe`.
> 
> I.e.
> 
> ```
>   - job_name: "mysql" # to get metrics  about mysql 
>     metrics_path: /probe
> ```
I want to use multi-target to avoid starting multiple processes on machine. my situation  ->
mysqld_exporter and prometheus version
```
root@m1:~# ls -l
total 89184
-rw-r--r-- 1 root root  7450991 Oct 18 18:03 mysqld_exporter-0.14.0.linux-amd64.tar.gz
-rw-r--r-- 1 root root 83865783 Oct 18 18:06 prometheus-2.37.1.linux-amd64.tar.gz
```

mysqld_exporter config
```
# /srv/metrics/mysqld_exporter/my.cnf
[client]
user = root
password = 123456
```

prometheus config
```
scrape_configs:
  # The job name is added as a label `job=<job_name>` to any timeseries scraped from this config.
  - job_name: "prometheus"

    # metrics_path defaults to '/metrics'
    # scheme defaults to 'http'.

    static_configs:
      - targets: ["localhost:9090"]

  - job_name: "mysqld_exporter"
    metrics_path: /probe
    static_configs:
      - targets:
        - 172.27.145.40:3306
        - 172.27.145.41:3306
    relabel_configs:
      - source_labels: [__address__]
        target_label: __param_target
      - source_labels: [__param_target]
        target_label: instance
      - target_label: __address__
        # The mysqld_exporter host:port
        replacement: localhost:9104
```

The prb is the targets is down
image

I don't know why this is，i think maybe the exporter verions is wrong? or something else?

**Comment by MattiasAng** [comment]
You need to compile it yourself, there is no release for the commits in this PR.

**Comment by shubhamsetia05** [comment]
@MattiasAng Any plans to release this (support to monitor multiple mysql DB's) in helm? We also have a use case to monitor multiple AWS RDS instances.
